File: src/shared/buri/aoc/data/path/Pathfinder.java
package buri.aoc.data.path;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import buri.aoc.data.tuple.BaseTuple;

/**
 * Utility class for search algorithms.
 * 
 * @author Brian Uri!
 */
public class Pathfinder {

	/**
	 * Constructs all paths from the start to the destinations and returns them with the shortest paths first.
	 */
	public static <T extends BaseTuple> List<Path> toPaths(T start, List<T> destinations, Map<T, T> cameFrom) {
		List<Path> paths = new ArrayList<>();
		for (T destination : destinations) {
			if (cameFrom.get(destination) != null) {
				paths.add(new Path(start, destination, cameFrom));
			}
		}
		Collections.sort(paths);
		return (paths);
	}
}

File: puzzles-kotlin/src/buri/aoc/common/Pathfinder.kt
package buri.aoc.common

import buri.aoc.common.position.Point2D

/**
 * Search utility for finding paths between points.
 *
 * @author Brian Uri!
 */
class Pathfinder(val stepStrategy: (Point2D<Int>) -> List<Point2D<Int>>) {

    /**
     * Returns a "came from" map showing all the reachable spaces from a particular space. The entry for the
     * starting position start will have a null value.
     */
    fun exploreFrom(start: Point2D<Int>): Map<Point2D<Int>, Point2D<Int>?> {
        val frontier = ArrayDeque<Point2D<Int>>()
        frontier.add(start)

        val cameFrom = mutableMapOf<Point2D<Int>, Point2D<Int>?>()
        cameFrom[start] = null

        var current: Point2D<Int>?
        while (frontier.isNotEmpty()) {
            current = frontier.removeFirst()
            for (next in stepStrategy(current).filter { !cameFrom.containsKey(it) }) {
                frontier.add(next)
                cameFrom[next] = current
            }
        }
        return cameFrom
    }
}

/**
 * Extension function to count the steps in a "came from" map. Returns -1 if there is no path between the points.
 */
fun <T> Map<T, T?>.countSteps(start: T, end: T): Int {
    if (start == end) {
        return 0
    }
    var steps = 0
    var current: T? = end
    while (current != null) {
        steps++
        current = this[current]
        if (current == start) {
            return steps
        }
    }
    return -1
}